modules/client/src/client-common/modules/ipfs.ts

```diff
@@ -72,30 +67,46 @@ export class IPFSModule implements IClientIpfsCore {
   // IPFS METHODS
 
   public async add(input: string | Uint8Array): Promise<string> {
-    return this.getOnlineClient().then(
-      (client) => client.add(input).then((res) => res.hash),
-    ).catch((e) => {
-      throw new Error(`Could not upload data: ${e?.message || ""}`);
-    });
+    return this.runAndRetryHelper(() => this.getClient().add(input)).then((
+      res,
+    ) => res.hash);
   }
 
-  public async pin(input: string): Promise<PinResponse> {
-    return this.getOnlineClient().then(
-      (client) => client.pin(input),
-    ).catch((e) => {
-      throw new Error(`Could not pin data: ${e?.message || ""}`);
-    });
+  public pin(input: string): Promise<PinResponse> {
+    return this.runAndRetryHelper(() => this.getClient().pin(input));
   }
 
   public fetchBytes(cid: string): Promise<Uint8Array | undefined> {
-    return this.getOnlineClient().then((client) => client.cat(cid));
+    return this.runAndRetryHelper(() => this.getClient().cat(cid));
+  }
+
+  private runAndRetryHelper<T>(f: () => Promise<T>): Promise<T> {
+    if (!this.clients.length) {
+      throw new ClientNotInitializedError("ipfs");
+    }
+    let retries = this.clients.length;
+    return runAndRetry({
+      func: f,
+      onFail: (e: Error) => {
+        if (e instanceof ClientError) {
+          if (e.response.status < 500) {
+            // If the error code is not a 5XX means the
+            // error is not generated by the server
+            throw new IpfsError();
+          }
+        }
+        retries--;
+        this.shiftClient();
+      },
+      shouldRetry: () => retries > 0,
+    });
```

Comment:
Why don't we recycle this code in a single place in common? We already have it for GraphQL